### PR TITLE
Reduce reliance on AtomicHelpers.h

### PR DESCRIPTION
**Pull request by ddn0**
AtomicHelpers provides some helpful functionality like atomic min and
atomic increment for floating point types (pre-C++20), but it also
contains reimplementations of STL functionality (std::fill,
std::transform) and basic C++ capabilities (variable assignment).

The main use of AtomicHelpers seems to be in supporting SyncStructures
which is itself mini-metaprogramming environment, and most of the
functionality in AtomicHelpers is not of general utility.

Remove various functions in AtomicHelpers in favor of their direct
implementation in SyncStructures.

Along the way, use fetch_add instead of compare_and_swap where we can.

### CODE REVIEW

**Comment by insertinterestingnamehere** [comment]
Related: https://github.com/IntelligentSoftwareSystems/Galois/pull/357. At some point it'd be nice to sync the repos again to minimize overlap of patches like this.

**Comment by ddn0** [comment]
Thanks for pointing on the overlap!

I've manually incorporated your changes here for now.

**Comment by insertinterestingnamehere** [comment]
Sounds good. Thanks!

**Review by tylershunt — APPROVED** [review]

**Review by gurbinder533 — APPROVED** [review]
LGTM. Much cleaner.

File: libgalois/include/galois/AtomicHelpers.h [hunk 1]
```diff
@@ -18,193 +18,74 @@
...
-const Ty min(std::atomic<Ty>& a, const Ty& b) {
-  Ty old_a = a;
-  if (a > b) {
-    a = b;
-  }
-  return old_a;
+#if __cplusplus > 201703L
+template <typename T>
+T atomicAdd(std::atomic<T>& a, const T& b) {
+  return a.fetch_add(b);
```
Comment (roshandathathri) [inline libgalois/include/galois/AtomicHelpers.h:50]: `a.fetch_add(b, std::memory_order_relaxed);`

File: libgalois/include/galois/AtomicHelpers.h [hunk 2]
```diff
@@ -18,193 +18,74 @@
...
-  Ty old_a = a;
-  if (a > b) {
-    a = b;
-  }
-  return old_a;
+#else
+template <typename T>
+T atomicAdd(std::atomic<T>& a, const T& b,
+            std::enable_if_t<std::is_integral_v<T>>* = nullptr) {
+  return a.fetch_add(b);
```
Comment (roshandathathri) [inline libgalois/include/galois/AtomicHelpers.h:56]: `a.fetch_add(b, std::memory_order_relaxed);`

File: libgalois/include/galois/AtomicHelpers.h [hunk 3]
```diff
@@ -18,193 +18,74 @@
...
 
-template <typename Ty>
-const Ty add(Ty& a, std::atomic<Ty>& b) {
-  Ty old_a = a;
-  a        = a + b.load();
-  return old_a;
+#if __cplusplus > 201703L
+template <typename T>
+T atomicSub(std::atomic<T>& a, const T& b) {
+  return a.fetch_sub(b);
```
Comment (roshandathathri) [inline libgalois/include/galois/AtomicHelpers.h:72]: `a.fetch_sub(b, std::memory_order_relaxed);`

File: libgalois/include/galois/AtomicHelpers.h [hunk 4]
```diff
@@ -18,193 +18,74 @@
...
-template <typename Ty>
-const Ty add(Ty& a, const Ty& b) {
-  Ty old_a = a;
-  a += b;
-  return old_a;
+#else
+template <typename T>
+T atomicSub(std::atomic<T>& a, const T& b,
+            std::enable_if_t<std::is_integral_v<T>>* = nullptr) {
+  return a.fetch_sub(b);
```
Comment (roshandathathri) [inline libgalois/include/galois/AtomicHelpers.h:78]: `a.fetch_sub(b, std::memory_order_relaxed);`

File: libgluon/include/galois/runtime/DataCommMode.h [hunk 5]
```diff
@@ -28,6 +28,8 @@
  */
 #pragma once
 
+#include <cstddef>
```
Comment (roshandathathri) [inline libgluon/include/galois/runtime/DataCommMode.h:31]: Why was this needed?

File: lonestar/analytics/distributed/sssp/sssp_pull.cpp [hunk 6]
```diff
@@ -27,6 +27,7 @@
 
 #include <iostream>
 #include <limits>
+#include <numeric>
```
Comment (roshandathathri) [inline lonestar/analytics/distributed/sssp/sssp_pull.cpp:30]: Why not algorithm?

File: libgluon/include/galois/runtime/SyncStructures.h [hunk 7]
```diff
@@ -446,7 +446,7 @@ class FieldFlags {
         return true;                                                           \
       }                                                                        \
       assert(personality == CPU);                                              \
-      galois::add(g.getData(node_id).fieldname, y);                            \
+      g.getData(node_id).fieldname = g.getData(node_id).fieldname + y;         \
```
Comment (roshandathathri) [inline libgluon/include/galois/runtime/SyncStructures.h:449]: I'm assuming you're not using `+=` because that operator might not have overloaded for that type? Forward thinking :)

**Review by roshandathathri — APPROVED** [review]
Yay!

File: libgluon/include/galois/runtime/DataCommMode.h (hunk 5 above)
Comment (ddn0) [reply libgluon/include/galois/runtime/DataCommMode.h:31]: to get the definition of `size_t`

File: libgluon/include/galois/runtime/SyncStructures.h (hunk 7 above)
Comment (ddn0) [reply libgluon/include/galois/runtime/SyncStructures.h:449]: yeah; this is actually true of `std::atomic` so I had to do it.

Ian's linked change does relaxed reads when the field is an std::atomic. I'll probably add that when I follow this PR with the refactor of SyncStructures to non static functions.

Comment (roshandathathri) [reply libgluon/include/galois/runtime/SyncStructures.h:449]: Cool.
